plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1890,6 +1895,46 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
         }
     }
 
+    private static void checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain<DeltaLakeColumnHandle> enforcedPartitionConstraint, IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<AddFileEntry> addedFiles = commitSummary.getIsBlindAppend().orElse(false)
+                ? List.of() // Do not conflict with blind appends
+                : commitSummary.getAddedFiles();
+
+        if (addedFiles.isEmpty()) {
+            return;
+        }
+
+        boolean readWholeTable = enforcedPartitionConstraint.isAll();
+        if (readWholeTable) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files added in the modified table by concurrent write operation.");
+        }
+
+        Map<DeltaLakeColumnHandle, Domain> enforcedDomains = enforcedPartitionConstraint.getDomains().orElseThrow();
+        Optional<AddFileEntry> unreadAddAction = addedFiles.stream()
+                .filter(addAction -> partitionMatchesPredicate(addAction.getCanonicalPartitionValues(), enforcedDomains))
```

Comment:
Should it match source enforceDomains with source addAction. Or with target?

Reply:
The reconciliation is done AFAIU based on the enforced partition constraints of the source tables (for non-blind INSERT or UPDATE/MERGE/DELETE operations).
If the operations writes new data - lNSERT operations - , this is not relevant for the WriteSerializable isolation level.
This aspect becomes relevant when dealing with Serializable operations on the same partition of the table - out of scope for the current PR.